File: TP's/TP 7/Padrao.java
import java.io.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;

class Tp4 {
    private String appID;
    private String name;
    private String releaseDate;
    private String compradores;
    private String preco;
    private String score;
    private String userScore;
    private String nConquistas;
    private String supportedLanguages;
    private String publishers;
    private String developers;
    private String categorias;
    private String generos;
    private String tags;

    public Tp4() {
    }

    public void setAppID(String appID) {
        this.appID = appID.trim();
    }

    public void setName(String name) {
        this.name = name.trim();
    }

    public void setReleaseDate(String releaseDate) {
        this.releaseDate = releaseDate.trim();
    }

    public void setCompradores(String compradores) {
        this.compradores = compradores.trim();
    }

    public void setPreco(String preco) {
        this.preco = preco.trim();
    }

    public void setScore(String score) {
        this.score = score.trim();
    }

    public void setUserScore(String userScore) {
        this.userScore = userScore.trim();
    }

    public void setnConquistas(String nConquistas) {
        this.nConquistas = nConquistas.trim();
    }

    public void setSupportedLanguages(String supportedLanguages) {
        this.supportedLanguages = supportedLanguages.trim();
    }

    public void setPublishers(String publishers) {
        this.publishers = publishers.trim();
    }

    public void setDevelopers(String developers) {
        this.developers = developers.trim();
    }

    public void setCategorias(String categorias) {
        this.categorias = categorias.trim();
    }

    public void setGeneros(String generos) {
        this.generos = generos.trim();
    }

    public void setTags(String tags) {
        this.tags = tags.trim();
    }

    public String getAppID() {
        return appID;
    }

    public String getPreco() {
        return preco;
    }

    @Override
    public String toString() {
        String dataFormatada = releaseDate;
        try {
            SimpleDateFormat entrada = new SimpleDateFormat("MMM dd, yyyy", Locale.ENGLISH);
            SimpleDateFormat saida = new SimpleDateFormat("dd/MM/yyyy");
            dataFormatada = saida.format(entrada.parse(releaseDate.replace("\"", "")));
        } catch (ParseException e) {
        }
        return "=> " + appID + " ## " +
                limpar(name) + " ## " +
                dataFormatada + " ## " +
                limpar(compradores) + " ## " +
                limpar(preco) + " ## " +
                "[" + limparLista(supportedLanguages) + "] ## " +
                limpar(score) + " ## " +
                limpar(userScore) + " ## " +
                limpar(nConquistas) + " ## " +
                "[" + limparLista(publishers) + "] ## " +
                "[" + limparLista(developers) + "] ## " +
                "[" + limparLista(categorias) + "] ## " +
                "[" + limparLista(generos) + "] ## " +
                "[" + limparLista(tags) + "] ##";
    }

    private String limpar(String campo) {
        if (campo == null)
            return "";
        return campo.replace("\"", "").replace("[", "").replace("]", "").trim();
    }

    private String limparLista(String campo) {
        campo = limpar(campo);
        String[] partes = campo.split(",");
        StringBuilder resultado = new StringBuilder();
        for (String parte : partes) {
            String item = parte.trim();
            if (!item.isEmpty()) {
                if (resultado.length() > 0)
                    resultado.append(", ");
                resultado.append(item);
            }
        }
        return resultado.toString();
    }
}

public class Padrao {
    public static void main(String[] args) {
        String path = "/tmp/games.csv";
        List<Tp4> jogos = new ArrayList<>();

        try (Scanner fileScanner = new Scanner(new File(path))) {
            if (fileScanner.hasNextLine())
                fileScanner.nextLine(); // pular cabeçalho
            while (fileScanner.hasNextLine()) {
                String linha = fileScanner.nextLine();
                String[] campos = linha.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)", -1);
                if (campos.length < 14)
                    continue;

                Tp4 jogo = new Tp4();
                jogo.setAppID(campos[0]);
                jogo.setName(campos[1]);
                jogo.setReleaseDate(campos[2]);
                jogo.setCompradores(campos[3]);
                jogo.setPreco(campos[4]);
                jogo.setSupportedLanguages(campos[5]);
                jogo.setScore(campos[6]);
                jogo.setUserScore(campos[7]);
                jogo.setnConquistas(campos[8]);
                jogo.setPublishers(campos[9]);
                jogo.setDevelopers(campos[10]);
                jogo.setCategorias(campos[11]);
                jogo.setGeneros(campos[12]);
                jogo.setTags(campos[13]);

                jogos.add(jogo);
            }
        } catch (FileNotFoundException e) {
            System.out.println("Arquivo não encontrado: " + e.getMessage());
            return;
        }

        Tp4[] arrayJogos = jogos.toArray(new Tp4[0]);
        Scanner input = new Scanner(System.in);
        String idBusca;

        while (true) {
            if (!input.hasNextLine())
                break;
            idBusca = input.nextLine().trim();
            if (idBusca.equalsIgnoreCase("Fim"))
                break;

            boolean encontrado = false;
            for (Tp4 jogo : arrayJogos) {
                if (idBusca.equals(jogo.getAppID())) {
                    System.out.println(jogo);
                    encontrado = true;
                    break;
                }
            }
            if (!encontrado)
                System.out.println("Jogo com ID " + idBusca + " não encontrado.");
        }

        input.close();
    }
}

File: TP's/TP 7/Padrao.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define TAMANHO_MAX_LINHA 4096
#define TAMANHO_MAX_ID 128
#define TAMANHO_MAX_CAMPO 1024
#define NUM_CAMPOS 14

typedef struct {
    char *appID;
    char *name;
    char *releaseDate;
    char *compradores;
    char *preco;
    char *score;
    char *userScore;
    char *nConquistas;
    char *supportedLanguages;
    char *publishers;
    char *developers;
    char *categorias;
    char *generos;
    char *tags;
} Tp4;

char* alocar_string(const char* s) {
    if (!s) return NULL;
    char *p = malloc(strlen(s) + 1);
    if (!p) return NULL;
    strcpy(p, s);
    return p;
}

void limpar(char* campo) {
    if (!campo) return;
    char *src = campo, *dst = campo;
    while (*src) {
        if (*src != '"' && *src != '[' && *src != ']') *dst++ = *src;
        src++;
    }
    *dst = '\0';
}

void formatarLista(char* campo) {
    if (!campo || !*campo) return;
    
    char temp[TAMANHO_MAX_CAMPO], resultado[TAMANHO_MAX_CAMPO] = "";
    strncpy(temp, campo, TAMANHO_MAX_CAMPO - 1);
    temp[TAMANHO_MAX_CAMPO - 1] = '\0';
    
    limpar(temp);

    char *token = strtok(temp, ",");
    bool primeiro = true;
    
    while (token) {
        while (*token == ' ') token++;
        
        if (!primeiro) strcat(resultado, ", ");
        strncat(resultado, token, TAMANHO_MAX_CAMPO - strlen(resultado) - 1);
        primeiro = false;
        token = strtok(NULL, ",");
    }
    
    strncpy(campo, resultado, TAMANHO_MAX_CAMPO - 1);
    campo[TAMANHO_MAX_CAMPO - 1] = '\0';
}

void formatarData(const char* entrada, char* saida) {
    const char* meses[] = {
        "Jan","Feb","Mar","Apr","May","Jun",
        "Jul","Aug","Sep","Oct","Nov","Dec"
    };
    
    int dia, ano;
    char mesStr[4];
    
    if (sscanf(entrada, "\"%3s %d, %d\"", mesStr, &dia, &ano) == 3) {
        int mes = 0;
        for (; mes < 12; mes++) {
            if (strcmp(mesStr, meses[mes]) == 0) break;
        }
        snprintf(saida, 20, "%02d/%02d/%04d", dia, mes + 1, ano);
    } else {
        strncpy(saida, entrada, 19);
        saida[19] = '\0';
    }
}

void imprimirTp4(Tp4 *jogo) {
    char dataFormatada[20];
    formatarData(jogo->releaseDate, dataFormatada);

    formatarLista(jogo->supportedLanguages);
    formatarLista(jogo->publishers);
    formatarLista(jogo->developers);
    formatarLista(jogo->categorias);
    formatarLista(jogo->generos);
    formatarLista(jogo->tags);

    limpar(jogo->name);
    limpar(jogo->compradores);
    limpar(jogo->preco);
    limpar(jogo->score);
    limpar(jogo->userScore);
    limpar(jogo->nConquistas);

    printf("=> %s ## %s ## %s ## %s ## %s ## [%s] ## %s ## %s ## %s ## [%s] ## [%s] ## [%s] ## [%s] ## [%s] ##\n",
           jogo->appID,
           jogo->name,
           dataFormatada,
           jogo->compradores,
           jogo->preco,
           jogo->supportedLanguages,
           jogo->score,
           jogo->userScore,
           jogo->nConquistas,
           jogo->publishers,
           jogo->developers,
           jogo->categorias,
           jogo->generos,
           jogo->tags);
}

void liberarTp4(Tp4 *jogo) {
    if (!jogo) return;
    
    free(jogo->appID);
    free(jogo->name);
    free(jogo->releaseDate);
    free(jogo->compradores);
    free(jogo->preco);
    free(jogo->score);
    free(jogo->userScore);
    free(jogo->nConquistas);
    free(jogo->supportedLanguages);
    free(jogo->publishers);
    free(jogo->developers);
    free(jogo->categorias);
    free(jogo->generos);
    free(jogo->tags);
    free(jogo);
}

Tp4* analisarLinha(char* linha) {
    Tp4 *jogo = calloc(1, sizeof(Tp4));
    if (!jogo) return NULL;

    char *campos[NUM_CAMPOS];
    char *token = strtok(linha, ",");
    int i = 0;
    
    while (token && i < NUM_CAMPOS) {
        campos[i++] = token;
        token = strtok(NULL, ",");
    }

    if (i < NUM_CAMPOS) {
        free(jogo);
        return NULL;
    }

    jogo->appID = alocar_string(campos[0]);
    jogo->name = alocar_string(campos[1]);
    jogo->releaseDate = alocar_string(campos[2]);
    jogo->compradores = alocar_string(campos[3]);
    jogo->preco = alocar_string(campos[4]);
    jogo->supportedLanguages = alocar_string(campos[5]);
    jogo->score = alocar_string(campos[6]);
    jogo->userScore = alocar_string(campos[7]);
    jogo->nConquistas = alocar_string(campos[8]);
    jogo->publishers = alocar_string(campos[9]);
    jogo->developers = alocar_string(campos[10]);
    jogo->categorias = alocar_string(campos[11]);
    jogo->generos = alocar_string(campos[12]);
    jogo->tags = alocar_string(campos[13]);

    return jogo;
}

bool buscarEImprimir(FILE *arquivo, const char *idBusca) {
    char linha[TAMANHO_MAX_LINHA];
    
    rewind(arquivo);
    fgets(linha, sizeof(linha), arquivo);
    
    while (fgets(linha, sizeof(linha), arquivo)) {
        Tp4 *jogo = analisarLinha(linha);
        if (!jogo) continue;
        
        if (strcmp(jogo->appID, idBusca) == 0) {
            imprimirTp4(jogo);
            liberarTp4(jogo);
            return true;
        }
        
        liberarTp4(jogo);
    }
    
    return false;
}

int main() {
    const char *caminho = "/tmp/games.csv";
    FILE *arquivo = fopen(caminho, "r");
    
    if (!arquivo) {
        fprintf(stderr, "Arquivo não encontrado.\n");
        return 1;
    }

    char linha[TAMANHO_MAX_LINHA];
    if (!fgets(linha, sizeof(linha), arquivo)) {
        fclose(arquivo);
        return 1;
    }

    char idBusca[TAMANHO_MAX_ID];
    while (fgets(idBusca, sizeof(idBusca), stdin)) {
        idBusca[strcspn(idBusca, "\n")] = 0;
        
        if (strcmp(idBusca, "Fim") == 0) break;
        
        buscarEImprimir(arquivo, idBusca);
    }

    fclose(arquivo);
    return 0;
}